client/rest/src/test/java/org/elasticsearch/client/RestClientGzipCompressionTests.java

```diff
@@ -0,0 +1,203 @@
+ * under the License.
+ */
+
+package org.elasticsearch.client;
+
+import com.sun.net.httpserver.HttpExchange;
+import com.sun.net.httpserver.HttpHandler;
+import com.sun.net.httpserver.HttpServer;
+import org.apache.http.HttpEntity;
+import org.apache.http.HttpHost;
+import org.apache.http.entity.ContentType;
+import org.apache.http.entity.StringEntity;
+import org.elasticsearch.mocksocket.MockHttpServer;
+import org.junit.AfterClass;
+import org.junit.Assert;
+import org.junit.BeforeClass;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.nio.charset.StandardCharsets;
+import java.util.concurrent.CompletableFuture;
+import java.util.zip.GZIPInputStream;
+import java.util.zip.GZIPOutputStream;
+
+public class RestClientGzipCompressionTests extends RestClientTestCase {
+
+    private static HttpServer httpServer;
+
+    @BeforeClass
+    public static void startHttpServer() throws Exception {
+        httpServer = MockHttpServer.createHttp(new InetSocketAddress(InetAddress.getLoopbackAddress(), 0), 0);
+        httpServer.createContext("/", new GzipResponseHandler());
+        httpServer.start();
+    }
+
+    @AfterClass
+    public static void stopHttpServers() throws IOException {
+        httpServer.stop(0);
+        httpServer = null;
+    }
+
+    /**
+     * A response handler that accepts gzip-encoded data and replies request and response encoding values
+     * followed by the request body. The response is compressed if "Accept-Encoding" is "gzip".
+     */
+    private static class GzipResponseHandler implements HttpHandler {
+        @Override
+        public void handle(HttpExchange exchange) throws IOException {
+
+            // Decode body (if any)
+            String contentEncoding = exchange.getRequestHeaders().getFirst("Content-Encoding");
+            InputStream body = exchange.getRequestBody();
+            if ("gzip".equals(contentEncoding)) {
+                body = new GZIPInputStream(body);
+            }
+            byte[] bytes = readAll(body);
+
+            boolean compress = "gzip".equals(exchange.getRequestHeaders().getFirst("Accept-Encoding"));
+            if (compress) {
+                exchange.getResponseHeaders().add("Content-Encoding", "gzip");
+            }
+
+            exchange.sendResponseHeaders(200, 0);
+
+            // Encode response if needed
+            OutputStream out = exchange.getResponseBody();
+            if (compress) {
+                out = new GZIPOutputStream(out);
+            }
+
+            // Outputs <request-encoding|null>#<response-encoding|null>#<request-body>
+            out.write(String.valueOf(contentEncoding).getBytes(StandardCharsets.UTF_8));
+            out.write('#');
+            out.write((compress ? "gzip" : "null").getBytes(StandardCharsets.UTF_8));
+            out.write('#');
```

Comment:
Is this custom encoding needed given the response headers will have the `Content-Encoding` header to be verified in the tests?

Reply:
Writing the request and response encoding in the response body is meant to allow the test to check what their values are from the server point of view, along with checking that both sides correctly encoded/decoded data if needed (i.e. both server and client were able to read the entity bodies).

Since from the caller's perspective all this is transparent and it receives an unencoded entity, we should expect `Content-Encoding` to be removed.

Testing further, I think the current behavior is buggy. When a gzip-encoded response is received:
* `response.getEntity().getContentEncoding()` is `null` (thanks to `GZIPDecompressingEntity`)
* `response.getHeader("Content-Encoding")` is `gzip`

As the entity presented to the application is no more encoded, both values should be `null`. We should therefore remove the response's `Content-Encoding` header in [`RestClient.convertResponse`](https://github.com/elastic/elasticsearch/blob/847a020e417320f6b1c02b816eda71ce79fe5fae/client/rest/src/main/java/org/elasticsearch/client/RestClient.java#L303-L308) where gzipped responses are decoded.

WDYT?

Reply:
That's a great point @swallez 

Would you like do add that support as part of this PR or open an issue and fix it separately?

Reply:
Let's focus on request compression in this PR. I've opened https://github.com/elastic/elasticsearch/issues/63314 for it.